V8 v12.3 patch

StefanStojanovic opened the pull request:
This patches V8 v12.3 for Windows, by fixing multiple compilation errors caused by V8 being a Clang-oriented project. There are various types of errors fixed by this going from changing `using` directives and renaming to overcoming the differences in which Clang and MSVC see templates and metaprogramming.

The changes introduced here are strictly meant as a patch only, so they shouldn't be pushed upstream.

Refs: https://github.com/targos/node/pull/13
Refs: https://github.com/targos/node/pull/14

StefanStojanovic commented:
@targos this patch doesn't land cleanly on v12.4. I'm working on it, but until it is ready, let's have this one as a backup. As a Node.js collaborator, can you run this through the CI with the `request-ci` label from here? If it cannot be done, I'll start the run manually.

targos commented:
> As a Node.js collaborator, can you run this through the CI with the request-ci label from here? If it cannot be done, I'll start the run manually.

I can't

StefanStojanovic committed:
deps: patch V8 to support compilation with MSVC

This patches V8 v12.3 for Windows, by fixing multiple compilation
errors caused by V8 being a Clang-oriented project. There are various
types of errors fixed by this going from changing `using` directives
and renaming to overcoming the differences in which Clang and MSVC see
templates and metaprogramming.

The changes introduced here are strictly meant as a patch only, so they
shouldn't be pushed upstream.

Refs: https://github.com/targos/node/pull/13
Refs: https://github.com/targos/node/pull/14

targos commented:
Could you post a link to the most recent CI run?

StefanStojanovic commented:
> Could you post a link to the most recent CI run?

Yes, here it is https://ci.nodejs.org/job/node-test-commit/69900/. It is a resumed build from https://ci.nodejs.org/job/node-test-commit/69886/.

From what I see, there are failures with `test-worker-arraybuffer-zerofill` which we've already seen in the V8 v12.2 update. There was https://github.com/nodejs/node/pull/51362/commits/f1881830e7d8a69806beea8ec95342ec73ee2c34 to address this, but it is not a part of the v12.3 update.

In addition, as seen in one of those 2 builds, but I also saw it previously when testing only Windows, `test-error-serdes` is showing signs of flakiness as well. We could mark that one as flaky too, if we're gonna mark the `test-worker-arraybuffer-zerofill` again.

Other than that I see no issues. I had some problems compiling on non-Windows platforms with my fixes, but I resolved that before running those 2 builds.

targos commented:
> There was https://github.com/nodejs/node/commit/f1881830e7d8a69806beea8ec95342ec73ee2c34 to address this, but it is not a part of the v12.3 update.

The change was in the `test` directory. It's still here: https://github.com/targos/node/blob/v8-123/test/parallel/parallel.status#L12-L13

targos commented:
Ok, then let's merge this, thank you!

targos commented on deps/v8/src/compiler/turboshaft/machine-optimization-reducer.h:
```diff
@@ -1349,53 +1349,23 @@ class MachineOptimizationReducer : public Next {
-              return __ Comparison(x, rhs, kind, rep_w);
-            }
-          }
+            CountLeadingSignBits(k2, rep_w) > k1 &&
+            matcher.Get(left).saturated_use_count.IsZero()) {
+          return __ Comparison(
+              x, __ WordConstant(base::bits::Unsigned(k2) << k1, rep_w), kind,
+              rep_w);
```
This change seems non-trivial. What's the effect of removing the `else if` branch?

StefanStojanovic replied:
That is a revert of [this commit](https://github.com/v8/v8/commit/8005bd282bf1565a11cae24a9811e99f8b3fc64b#diff-f11fdde01fe8fbd8c3fa045a3aadf1d6b06d0d442a6e6336626cd0b4361cced0). It was also a part of [v12.2 fix](https://github.com/nodejs/node/commit/b9d806a2dd7d6b137cd9f0099ec0f8ed0abea4d6#diff-ba5f45f27bb33b78f2da3ec5ade73dd18f61046810dbce4cca7ecad133fe7504) before.
